File: src/python/xerxes/extensions/skill_authoring/verifier.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Skill verification recipe generation and evaluation.

``SkillVerifier`` produces ``VerificationStep`` objects from a
``SkillCandidate`` and can later validate whether another candidate satisfies
those steps.
"""

from __future__ import annotations

import typing as tp
from dataclasses import dataclass, field

from .tracker import SkillCandidate, ToolCallEvent


@dataclass
class VerificationStep:
    """A single assertion in a skill verification recipe.

    Attributes:
        kind: Step type (e.g. ``call_count``, ``sequence_position``, ``args_subset``).
        tool_name: Target tool for sequence and args checks.
        position: Expected index in the observed event list, when applicable.
        args_required: Argument keys (and optional values) that must be present.
        expected_count: Expected number of successful calls for ``call_count`` checks.
        message: Human-readable description for logs and reports.
    """

    kind: str
    tool_name: str = ""
    position: int | None = None
    args_required: dict[str, tp.Any] = field(default_factory=dict)
    expected_count: int | None = None
    message: str = ""


@dataclass
class VerificationResult:
    """Outcome of verifying a candidate against a recipe.

    Attributes:
        passed: True when every step passed.
        passed_steps: Indices of steps that passed.
        failed_steps: ``(index, reason)`` pairs for steps that failed.
    """

    passed: bool
    passed_steps: list[int] = field(default_factory=list)
    failed_steps: list[tuple[int, str]] = field(default_factory=list)


class SkillVerifier:
    """Generate and evaluate skill verification recipes."""

    def generate(self, candidate: SkillCandidate) -> list[VerificationStep]:
        """Build a recipe from ``candidate.successful_events``."""

        steps: list[VerificationStep] = []
        successful = candidate.successful_events
        steps.append(
            VerificationStep(
                kind="call_count",
                expected_count=len(successful),
                message=f"expects {len(successful)} successful tool calls",
            )
        )
        for i, ev in enumerate(successful):
            steps.append(
                VerificationStep(
                    kind="sequence_position",
                    tool_name=ev.tool_name,
                    position=i,
                    message=f"position {i} should call {ev.tool_name}",
                )
            )
            if ev.arguments:
                steps.append(
                    VerificationStep(
                        kind="args_subset",
                        tool_name=ev.tool_name,
                        position=i,
                        args_required={k: ev.arguments[k] for k in list(ev.arguments)[:3]},
                        message=f"{ev.tool_name} expects keys {list(ev.arguments)[:3]}",
                    )
                )
        return steps

    def verify(
        self,
        steps: list[VerificationStep],
        candidate: SkillCandidate,
    ) -> VerificationResult:
        """Evaluate every ``step`` against ``candidate.successful_events``."""

        observed = candidate.successful_events
        passed: list[int] = []
        failed: list[tuple[int, str]] = []
        for i, step in enumerate(steps):
            ok, reason = self._evaluate(step, observed)
            if ok:
                passed.append(i)
            else:
                failed.append((i, reason))
        return VerificationResult(passed=not failed, passed_steps=passed, failed_steps=failed)

    def _evaluate(
        self,
        step: VerificationStep,
        observed: list[ToolCallEvent],
    ) -> tuple[bool, str]:
        """Return ``(True, "")`` if ``step`` passes against ``observed``, else ``(False, reason)``."""

        if step.kind == "call_count":
            if step.expected_count is not None and len(observed) != step.expected_count:
                return False, f"expected {step.expected_count} successful calls, got {len(observed)}"
            return True, ""
        if step.kind == "tool_called":
            if not any(e.tool_name == step.tool_name for e in observed):
                return False, f"tool {step.tool_name!r} was never called"
            return True, ""
        if step.kind == "sequence_position":
            if step.position is None or step.position >= len(observed):
                return False, f"position {step.position} not in observed sequence"
            actual = observed[step.position].tool_name
            if actual != step.tool_name:
                return False, f"expected {step.tool_name!r} at pos {step.position}, got {actual!r}"
            return True, ""
        if step.kind == "args_subset":
            if step.position is None or step.position >= len(observed):
                return False, "position out of range"
            actual_args = observed[step.position].arguments
            for k in step.args_required:
                if k not in actual_args:
                    return False, f"missing required arg key {k!r}"
            return True, ""
        if step.kind == "status_success":
            if step.position is None or step.position >= len(observed):
                return False, "position out of range"
            return (observed[step.position].status == "success"), "non-success status"
        return False, f"unknown step kind {step.kind!r}"

File: src/typescript/src/extensions/skillAuthoring/verifier.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import type { SkillCandidate, ToolArguments, ToolCallEvent } from './model.js'

export type VerificationStepKind = 'args_subset' | 'call_count' | 'sequence_position' | 'status_success' | 'tool_called'

/** One transparent assertion in a repeatable skill verification recipe. */
export interface VerificationStep {
  readonly argsRequired?: ToolArguments
  readonly expectedCount?: number
  readonly kind: VerificationStepKind | string
  readonly message: string
  readonly position?: number
  readonly toolName?: string
}

export interface VerificationFailure {
  readonly index: number
  readonly reason: string
}

export interface VerificationResult {
  readonly failedSteps: readonly VerificationFailure[]
  readonly passed: boolean
  readonly passedSteps: readonly number[]
}

export interface SkillVerifierOptions {
  /** Preserve Python-compatible key-only argument checks unless hosts explicitly opt into value equality. */
  readonly compareArgumentValues?: boolean
}

/** Builds and evaluates deterministic recipes from observed successful calls. */
export class SkillVerifier {
  private readonly compareArgumentValues: boolean

  constructor(options: SkillVerifierOptions = {}) {
    this.compareArgumentValues = options.compareArgumentValues ?? false
  }

  generate(candidate: SkillCandidate): VerificationStep[] {
    const successful = candidate.successfulEvents
    const steps: VerificationStep[] = [{
      kind: 'call_count',
      expectedCount: successful.length,
      message: 'expects ' + successful.length + ' successful tool calls',
    }]
    for (let position = 0; position < successful.length; position += 1) {
      const event = successful[position]
      if (!event) {
        continue
      }
      steps.push({
        kind: 'sequence_position',
        toolName: event.toolName,
        position,
        message: 'position ' + position + ' should call ' + event.toolName,
      })
      const arguments_ = firstArguments(event.arguments, 3)
      if (Object.keys(arguments_).length) {
        steps.push({
          kind: 'args_subset',
          toolName: event.toolName,
          position,
          argsRequired: arguments_,
          message: event.toolName + ' expects keys ' + Object.keys(arguments_).join(', '),
        })
      }
    }
    return steps
  }

  verify(steps: readonly VerificationStep[], candidate: SkillCandidate): VerificationResult {
    const observed = candidate.successfulEvents
    const passedSteps: number[] = []
    const failedSteps: VerificationFailure[] = []
    for (let index = 0; index < steps.length; index += 1) {
      const step = steps[index]
      if (!step) {
        continue
      }
      const result = this.evaluateStep(step, observed)
      if (result.ok) {
        passedSteps.push(index)
      } else {
        failedSteps.push({ index, reason: result.reason })
      }
    }
    return { passed: failedSteps.length === 0, passedSteps, failedSteps }
  }

  private evaluateStep(
    step: VerificationStep,
    observed: readonly ToolCallEvent[],
  ): { readonly ok: boolean; readonly reason: string } {
    switch (step.kind) {
      case 'call_count':
        if (step.expectedCount !== undefined && observed.length !== step.expectedCount) {
          return failure('expected ' + step.expectedCount + ' successful calls, got ' + observed.length)
        }
        return success()
      case 'tool_called':
        if (!step.toolName || !observed.some(event => event.toolName === step.toolName)) {
          return failure('tool ' + String(step.toolName) + ' was never called')
        }
        return success()
      case 'sequence_position': {
        const event = eventAt(step.position, observed)
        if (!event) {
          return failure('position ' + String(step.position) + ' not in observed sequence')
        }
        if (event.toolName !== step.toolName) {
          return failure('expected ' + String(step.toolName) + ' at pos ' + step.position + ', got ' + event.toolName)
        }
        return success()
      }
      case 'args_subset': {
        const event = eventAt(step.position, observed)
        if (!event) {
          return failure('position out of range')
        }
        const required = step.argsRequired ?? {}
        for (const [key, value] of Object.entries(required)) {
          if (!Object.hasOwn(event.arguments, key)) {
            return failure('missing required arg key ' + key)
          }
          if (this.compareArgumentValues && !valuesEqual(event.arguments[key], value)) {
            return failure('argument value did not match for key ' + key)
          }
        }
        return success()
      }
      case 'status_success': {
        const event = eventAt(step.position, observed)
        if (!event) {
          return failure('position out of range')
        }
        return event.status === 'success' ? success() : failure('non-success status')
      }
      default:
        return failure('unknown step kind ' + step.kind)
    }
  }
}

function eventAt(position: number | undefined, events: readonly ToolCallEvent[]): ToolCallEvent | undefined {
  if (position === undefined || !Number.isInteger(position) || position < 0) {
    return undefined
  }
  return events[position]
}

function firstArguments(arguments_: ToolArguments, count: number): ToolArguments {
  const result: Record<string, unknown> = {}
  for (const key of Object.keys(arguments_).slice(0, count)) {
    result[key] = arguments_[key]
  }
  return result
}

function valuesEqual(left: unknown, right: unknown): boolean {
  if (Object.is(left, right)) {
    return true
  }
  if (Array.isArray(left) && Array.isArray(right)) {
    return left.length === right.length && left.every((value, index) => valuesEqual(value, right[index]))
  }
  if (isRecord(left) && isRecord(right)) {
    const leftKeys = Object.keys(left).sort()
    const rightKeys = Object.keys(right).sort()
    return leftKeys.length === rightKeys.length
      && leftKeys.every((key, index) => key === rightKeys[index] && valuesEqual(left[key], right[key]))
  }
  return false
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

function failure(reason: string): { readonly ok: false; readonly reason: string } {
  return { ok: false, reason }
}

function success(): { readonly ok: true; readonly reason: string } {
  return { ok: true, reason: '' }
}
